# Assignment 3

Please complete the following three tasks on the dataset assigned to you. Inspect the dataset visually and using proper Python instrumentation, and establish whether:

* the dataset is tidy? If yes, demonstrate (Python code) and describe why this is the case. If no, reshape (Python code) the dataset into a tidy one and describe the steps performed.
* the dataset includes missing data? If no, demonstrate (Python code) and describe why this conclusion is warranted. If yes, apply a handling strategy for missing data (Python code) and discuss the consequences.
* the dataset includes duplicate data? If no, demonstrate (Python code) and describe why this conclusion is justified. If yes, deduplicate the data set (Python code) and describe the steps performed.

## Task 1 (8 credits): Tidy vs. messy?

* Is the dataset tidy according to the three rules covered in Unit 3?
* If yes, show and discuss why?
* If no, tidy the dataset and document the steps performed?

In addition, characterize the final, tidy dataset:
* What are the data objects? How many different kinds of data objects are covered?
* What are the different variables? Describe each variable along the dimensions covered in Unit 3.

In [1]:
#import pandas library
import pandas as pd

#read dataset in the variable df
df = pd.read_csv('./data/data_notebook-1_pisascience.csv')

#print shape:(rows, columns) 
print('Shape:' + str(df.shape), "\n")

#sort data by Location and within location by time
df.sort_values(['LOCATION', 'TIME'], ascending=True, inplace=True)
#indexes got mixed by sorting the data
#therefore reset index 
df.reset_index(inplace=True)
#drop index column which is created automatically 
df = df.drop(columns='index')

#convert the last two column names to upper cases to match the other column names
df = df.rename(columns = {"Value":"VALUE", "Flag Codes": "FLAG CODES"})

#print the different types of attributes
print(df.dtypes, "\n")

#convert last two column names to upper cases to match the other columns
df = df.rename(columns = {"Value":"VALUE", "Flag Codes": "FLAG CODES"})

display(df.head())

#raise NotImplementedError()

Shape:(474, 8) 

LOCATION       object
INDICATOR      object
SUBJECT        object
MEASURE        object
FREQUENCY      object
TIME            int64
VALUE         float64
FLAG CODES    float64
dtype: object 



LOCATION    INDICATOR SUBJECT    MEASURE FREQUENCY  TIME  VALUE  FLAG CODES
0      AUS  PISASCIENCE     BOY  MEANSCORE         A  2006  527.0         NaN
1      AUS  PISASCIENCE    GIRL  MEANSCORE         A  2006  527.0         NaN
2      AUS  PISASCIENCE     TOT  MEANSCORE         A  2006  527.0         NaN
3      AUS  PISASCIENCE     BOY  MEANSCORE         A  2009  527.0         NaN
4      AUS  PISASCIENCE    GIRL  MEANSCORE         A  2009  528.0         NaN

## Tidy or messy dataset?
Those 3 rules which have to be satisfied if dataset is tidy:

####  Each variable forms a column.
* This statement is satisfied. 
* Explanation: Each variable forms a column. The columns are variable by variable. Labels (Headers) denote variable names, not values. There is always just one variable stored in one column. 

#### Each observation forms a row.
* This statement is satisfied. 
* Explanation: the rows are observation by observation. Every measurement forms a different row. 

#### Each type of data object (observation unit) forms a separate table.
* This statement is satisfied. 
* Explanation: Each value has its own cell and as said before there is always just one variable stored in one column.

## Characterize dataset

####  Structure
* This dataset has 474 rows and 8 columns as you can see with the function df.shape. 
* Therefore, the dataset could theoretically contain 3792 (474x8) different values. The last column "Flag Codes" doesn't contain any values, so it's actually containing 3318 (474x7) values. 
* I have sorted the dataframe with the df.sort_values() function. First I sorted by Location and within the locations I sorted by year. Afterwards I had to reset my indexes, because there got mixed up while sorting the data.
* Another thing I've done to make my dataset more visually appealing was converting all column names to upper cases.
* As mentioned before the column "Flag Codes" doesn't contain any values, this content will be handled in Task 2.


####  Objects
* After having the sorted dataset, you can see at first sight that the observation are measuring the meanscore for BOYS, GIRLS and TOTAL in the years 2006, 2009, 2012, 2015. 
* For example, AUS has three observations concerning 2006, such as three observations concerning 2009 and so on. Therefore, you can find 12 objects with AUS as Location. 
* There are some locations which are exceptions, but this will be discussed in Task 2. 

####  Variables
I have determined the types of the different variables with df.dtypes. 

* LOCATION:    object; categorical; nominal scale; fixed variables
* INDICATOR:   object; categorical; nominal scale; fixed variables
* SUBJECT:     object; categorical; nominal scale; fixed variables
* MEASURE:     object; categorical; nominal scale; fixed variables
* FREQUENCY:   object; categorical; nominal scale; fixed variables
* TIME:        int64; numeric; interval scale; fixed variables
* Value:       float64; numeric; ratio scale; measured variables
* Flag Codes:  the function states that the flag codes are float64, but we don't have any values in there so I can't analyze the type any further

## Task 2 (4 credits): Missing data?

* Does the dataset contain missing data?
* If no, show and give proof that this is actually the case.
* If yes, extract the missings and describe them.
* If yes, apply a (simple) handling strategy and briefly discuss the implications.

In [2]:
#print shape
print('Shape:' + str(df.shape), "\n")

#print all variables and count the many missing values 
print(df.isna().sum(), "\n") 

#print only comlumn 'FLAG CODES'
#every value in column 'FLAG CODES' is missing 
df['FLAG CODES'].head(10)

#drop the column 'FLAG CODES'
df2=df.drop(columns='FLAG CODES')

#print the cleaned data without column 'Flag Codes'
display(df2.head())



#count how many unique values are located in LOCATION
print(str(len(df2['LOCATION'].value_counts()))+' different Values in LOCATION', "\n")
#how many measurements in each location
df2['LOCATION'].value_counts()

#count values of the attribute Subject 
df2['SUBJECT'].value_counts()

#create new variable for all rows with the subject TOT 
df_subject_tot=df2[df2['SUBJECT']=='TOT']
#create new variable for all rows with the subject GIRL 
df_subject_girl=df2[df2['SUBJECT']=='GIRL']
#create new variable for all rows with the subject BOY 
df_subject_boy=df2[df2['SUBJECT']=='BOY']

#count how many times TOT appears per year
print('Number of Measurements for each Year with Subject TOT:')
print(df_subject_tot['TIME'].value_counts(), "\n")
#count how many times GIRL appears per year
print('Number of Measurements for each Year with Subject GIRL:')
print(df_subject_girl['TIME'].value_counts(), "\n")
#count how many times BOY appears per year
print('Number of Measurements for each Year with Subject BOY:')
print(df_subject_boy['TIME'].value_counts(), "\n")
#2015: TOT one measure less, 2012: TOT one measure more, 2009: TOT one measure more, 2006: TOT two measures more

#rows with location LVA
#LVA: just one Measure (TOT) in 2006,2009,2012, two Measures (Boy, Girl) in 2015
df_location_lva=df2[df2['LOCATION']=='LVA'] 

#rows with location RUS
#RUS: just one Measure (TOT) in 2006, two Measures (Boy, Girl)in 2015
df_location_rus=df2[df2['LOCATION']=='RUS'] 

#pivot dataframe for handling missing values easier
df3 = df2.set_index(['LOCATION','INDICATOR','SUBJECT','MEASURE','FREQUENCY','TIME'])['VALUE'].unstack().reset_index()
df_location_lva=df3.loc[df3['LOCATION']=='LVA'] 
display(df_location_lva)

#fill missing values withe the column mean
df3 = df3.bfill(axis=1).ffill(axis=1)

#print location RUS and LVA with imputated mean
df_location_rus=df3.loc[df3['LOCATION']=='RUS'] 
df_location_lva=df3.loc[df3['LOCATION']=='LVA'] 
display(df_location_lva)

#Using the .melt() function to reshape the data set again
df4= pd.melt(df3, id_vars=['LOCATION','INDICATOR','SUBJECT','MEASURE','FREQUENCY'], value_vars=[2006,2009,2012,2015],
var_name='TIME', value_name='Value')

#sort dataframe again 
df4.sort_values(['LOCATION', 'TIME'], ascending=True, inplace=True)
df4.reset_index(inplace=True)
#drop index column which is created automatically 
df4.drop(columns='index', inplace=True)

#looking at the processed data set
df4.head()

#checking locations e.g. LVA and RUS to check if missing rows were imputed
df_location_rus=df4.loc[df4['LOCATION']=='RUS'] 
df_location_lva=df4.loc[df4['LOCATION']=='LVA'] 
display(df_location_lva)

#showing that each location contains 12 rows
df4["LOCATION"].value_counts()
#showing that each subject contains 180 rows
print(df4['SUBJECT'].value_counts(), "\n")

#print new shape
print('Shape:' + str(df4.shape))

#raise NotImplementedError()

Shape:(474, 8) 

LOCATION        0
INDICATOR       0
SUBJECT         0
MEASURE         0
FREQUENCY       0
TIME            0
VALUE           0
FLAG CODES    474
dtype: int64 



LOCATION    INDICATOR SUBJECT    MEASURE FREQUENCY  TIME  VALUE
0      AUS  PISASCIENCE     BOY  MEANSCORE         A  2006  527.0
1      AUS  PISASCIENCE    GIRL  MEANSCORE         A  2006  527.0
2      AUS  PISASCIENCE     TOT  MEANSCORE         A  2006  527.0
3      AUS  PISASCIENCE     BOY  MEANSCORE         A  2009  527.0
4      AUS  PISASCIENCE    GIRL  MEANSCORE         A  2009  528.0

45 different Values in LOCATION 

Number of Measurements for each Year with Subject TOT:
2015    44
2012    39
2006    39
2009    38
Name: TIME, dtype: int64 

Number of Measurements for each Year with Subject GIRL:
2015    45
2012    38
2009    37
2006    37
Name: TIME, dtype: int64 

Number of Measurements for each Year with Subject BOY:
2015    45
2012    38
2009    37
2006    37
Name: TIME, dtype: int64 



TIME LOCATION    INDICATOR SUBJECT    MEASURE FREQUENCY   2006   2009   2012  \
81        LVA  PISASCIENCE     BOY  MEANSCORE         A    NaN    NaN    NaN   
82        LVA  PISASCIENCE    GIRL  MEANSCORE         A    NaN    NaN    NaN   
83        LVA  PISASCIENCE     TOT  MEANSCORE         A  490.0  494.0  502.0   

TIME   2015  
81    485.0  
82    496.0  
83    490.0

TIME LOCATION    INDICATOR SUBJECT    MEASURE FREQUENCY   2006   2009   2012  \
81        LVA  PISASCIENCE     BOY  MEANSCORE         A  485.0  485.0  485.0   
82        LVA  PISASCIENCE    GIRL  MEANSCORE         A  496.0  496.0  496.0   
83        LVA  PISASCIENCE     TOT  MEANSCORE         A  490.0  494.0  502.0   

TIME   2015  
81    485.0  
82    496.0  
83    490.0

LOCATION    INDICATOR SUBJECT    MEASURE FREQUENCY  TIME  Value
324      LVA  PISASCIENCE     BOY  MEANSCORE         A  2006  485.0
325      LVA  PISASCIENCE    GIRL  MEANSCORE         A  2006  496.0
326      LVA  PISASCIENCE     TOT  MEANSCORE         A  2006  490.0
327      LVA  PISASCIENCE     BOY  MEANSCORE         A  2009  485.0
328      LVA  PISASCIENCE    GIRL  MEANSCORE         A  2009  496.0
329      LVA  PISASCIENCE     TOT  MEANSCORE         A  2009  494.0
330      LVA  PISASCIENCE     BOY  MEANSCORE         A  2012  485.0
331      LVA  PISASCIENCE    GIRL  MEANSCORE         A  2012  496.0
332      LVA  PISASCIENCE     TOT  MEANSCORE         A  2012  502.0
333      LVA  PISASCIENCE     BOY  MEANSCORE         A  2015  485.0
334      LVA  PISASCIENCE    GIRL  MEANSCORE         A  2015  496.0
335      LVA  PISASCIENCE     TOT  MEANSCORE         A  2015  490.0

TOT     180
GIRL    180
BOY     180
Name: SUBJECT, dtype: int64 

Shape:(540, 7)


##  Missing data

### Column: Flag Codes
For determining any missing values I've used the df.isnull().sum() function. As you can see above it shows how many missing values every column has. The result is very clear; there are no missing values in the first 7 columns, but the last column doesn't contain any value at all.

#### Why are the Flag Codes missing 
Flag Codes Definition: "Flags represent metadata that recurs frequently within a dataset.
Flags appear as letters in parentheses in the cell alongside the data figures to which they correspond."(Source:stats.oecd.org)


Knowing what Flag Codes are, we can interpret the missing values as <b>Missing at random (MAR)</b>!
Explanation: The probability of a value being missing is associated with other variables in the data set (is unrelated to the value itself). As already stated in the definition flags are corresponding to other data figures.

#### Handling missing Flag Code Values
Because of the fact that the column Flag Codes doesn't have any value, imputation (mean or mode) wouldn't work in this case. In my opinion the best way how to handle this column is Deletion. I've dropped the whole column "FLAG CODES" because it doesn't provide any information at all and could be seen as useless in this case. I've stored the new dataframe without the last column in a new variable "df2".

### Column: Subject & Time 
While I was inspecting my sorted Dataframe, I've noticed that there is some inconstancy within the column 'SUBJECT'. As I explained Task 1 there should be 3 Subject values for each of the 4 years, so in total 12 values per year and location. As we have 45 different locations, we should have 45 rows for each different subject (TOT, GIRL, BOYS). But this isn't the case. 

By printing the created variables (df_subject_tot, df_subject_girl, df_subject_boy) one can notice that we're missing 66 objects (rows):
* 1 'TOT' in 2015, 6 'TOT' in 2012 and 2006 and 7 'TOT' in 2009 
* 7  'GIRL'/'BOY' in 2012, 8 'GIRL'/'BOY' in 2009 and 2006

#### Why are some Subjects missing 
Some locations don't provide the meanscore for every year and every Subject. The function df2['LOCATION'].value_counts() gave an overview about how many different measurements are stored in each location. This showed that the locations RUS, AUT contain 9 rows instead of 12, the location LVA only provides 6 rows and HKG, PER, TWN, COL, MAC, SGP just 3. 
Reasons for missingness: Not all variables (knowingly) available at all points in time of data recording (measurement).

#### Handling missing Subjects
To fill the missing measurements, I've decided to performe imputation. I pivot the dataframe to get the TIME-values as seperate columns. Consequently we get NaN for the missing meanscores for each subject in each year. I've replaced the missing values by imputing the value located to the right/left of the missing value with the functions  bfill() and  ffill(). Afterwards I use the pd.melt (the process is printed for the location LVA as an example).
To check my result, I've used the function df4['SUBJECT'].value_counts() which gives the counted values of each Subject. TOT, GIRL and BOY finally contain 180 values each (4 values for each year multiplied by 45 countries).

* Shape before: (474, 8) 
* Shape now: (540, 7)
* Rows increased by 66 (impute missing objects)
* Columns decreased by 1 (drop 'FLAG CODE' column)

## Task 3 (3 credits): Duplicate data?

* Does the dataset contain duplicates?
* If no, show and give proof that this is actually the case.
* If yes, extract the duplicates and describe them.
* If yes, make an attempt to remove duplicates.

In [3]:
#check if dataset has duplicates 
duplicates=df4.duplicated().sum()
#print duplicates
print('Duplicates: ' + str(duplicates))

#raise NotImplementedError()

Duplicates: 0


## No Duplicate Data


Luckily there are no duplicates in my dataset!

### Proof
* The function df.duplicated().sum() gives me the number of duplicates in the dataframe. 
* Another way of proofing that my dataframe doesn't contain any duplicates is handling the missing rows in Task 2. I've clarified that each location contains 12 objects. If we would have any duplicates I would have noticed while inspecting and imputing the missing rows.